enterprise/server/githubapp/githubapp.go

```diff
@@ -586,6 +592,98 @@ func (a *GitHubApp) GetAccessibleGitHubRepos(ctx context.Context, req *ghpb.GetA
 	return &ghpb.GetAccessibleReposResponse{RepoUrls: urls}, nil
 }
 
+func (a *GitHubApp) CreateRepo(ctx context.Context, req *repb.CreateRepoRequest) (*repb.CreateRepoResponse, error) {
+	tu, err := a.env.GetUserDB().GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if tu.GithubToken == "" {
+		return nil, status.UnauthenticatedError("github account link is required")
+	}
+
+	// Pick the right client based on the request (organization or user).
+	githubClient, token, err := a.newAuthenticatedClient(ctx, tu.GithubToken)
+	if req.Organization != "" {
+		githubClient, token, err = a.newInstallationClient(ctx, req.InstallationId)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new repository on github
+	repo, _, err := githubClient.Repositories.Create(ctx, req.Organization, &github.Repository{
+		Name:        github.String(req.Name),
+		Description: github.String(req.Description),
+		Private:     github.Bool(req.Private),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// If we have a template, copy the template contents to the new repo
+	if req.Template != "" {
+		tmpDir := fmt.Sprintf("%s/template-repo/%d/%s/", os.TempDir(), req.InstallationId, req.Name)
+		cloneTemplate(tmpDir, token, req.Template, *repo.CloneURL)
+	}
+
+	// Link new repository to enable workflows
+	wfs := a.env.GetWorkflowService()
+	if wfs == nil {
+		return nil, status.UnimplementedErrorf("no workflow service configured")
+	}
+	_, err = a.LinkGitHubRepo(ctx, &ghpb.LinkRepoRequest{
+		RequestContext: req.RequestContext,
+		RepoUrl:        *repo.HTMLURL,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &repb.CreateRepoResponse{
+		RepoUrl: *repo.HTMLURL,
+	}, nil
+}
+
+func cloneTemplate(tmpDir, token, srcURL, destURL string) error {
+	// Make a temporary directory for the template
+	err := os.MkdirAll(tmpDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	dir, err := ioutil.TempDir(tmpDir, "repo")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(dir)
+
+	// Clone the template into the directory
+	auth := &githttp.BasicAuth{
+		Username: "github",
+		Password: token,
+	}
+	gitRepo, err := git.PlainClone(dir, false, &git.CloneOptions{
+		URL:  srcURL,
+		Auth: auth,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Create a new remote and push to it
+	remote, err := gitRepo.CreateRemote(&config.RemoteConfig{
+		Name: "new-repo",
```

Comment:
should this be 'origin' or something? I'm a bit ig when it comes to git

Reply:
a remote called `origin` already exists (that's the original template repo that we cloned)

i originally had this deleting `origin` and recreating this new remote as `origin`, but figured it was cleaner / less steps to just create a new remote with a new name and push to that one